src/ops/mod.rs

```diff
@@ -9,92 +9,136 @@ pub mod upgrade;
 
-impl ExitError {
-    /// Exit 1 with an exit message
-    pub fn errmsg<T>(message: T) -> ExitError
+    /// Return an OpResult with a final message to print before exit 0
+    /// Note, the final message is possibly intended to be consumed
+    /// by automated tests.
+    pub fn ok_msg<T>(message: T) -> OpResult
     where
         T: Into<String>,
     {
-        ExitError::err(1, message.into())
+        Ok(Some(message.into()))
     }
 
-    /// Helpers to create exit results
-    ///
-    /// Note: err panics if exitcode is zero.
-    fn err<T>(exitcode: i32, message: T) -> ExitError
-    where
-        T: Into<String>,
-    {
-        assert!(exitcode != 0, "ExitError exitcode must be > 0!");
+    /// Return an OpResult with no message to be printed, producing
+    /// a silent exit 0
+    pub fn ok() -> OpResult {
+        Ok(None)
+    }
 
-        ExitError {
-            exitcode,
-            message: message.into(),
+    impl ExitError {
+        /// Exit 1 to signify a generic expected error
+        /// (e.g. something that sometimes just goes wrong, like a nix build).
+        pub fn expected_error<T>(message: T) -> ExitError
+        where
+            T: Into<String>,
+        {
+            ExitError {
+                exitcode: 1,
+                message: message.into(),
+            }
         }
-    }
 
-    /// Exit code of the failure message, guaranteed to be > 0
-    pub fn exitcode(&self) -> i32 {
-        self.exitcode
-    }
+        /// Exit 100 to signify a user error (“the user is holding it wrong”).
+        /// This is a permanent error, if the program is executed the same way
+        /// it should crash with 100 again.
+        pub fn user_error<T>(message: T) -> ExitError
+        where
+            T: Into<String>,
+        {
+            ExitError {
+                exitcode: 100,
+                message: message.into(),
+            }
+        }
 
-    /// Exit message to be displayed to the user on stderr
-    pub fn message(&self) -> &str {
-        &self.message
-    }
-}
+        /// Exit 101 to signify an unexpected crash (failing assertion or panic).
+        /// This is the same exit code that `panic!()` emits.
+        pub fn panic<T>(message: T) -> ExitError
+        where
+            T: Into<String>,
+        {
+            ExitError {
+                exitcode: 101,
+                message: message.into(),
+            }
+        }
 
-#[cfg(test)]
-mod tests {
-    use super::ExitError;
+        /// Exit 111 to signify a temporary error (such as resource exhaustion)
+        pub fn temporary<T>(message: T) -> ExitError
+        where
+            T: Into<String>,
+        {
+            ExitError {
+                exitcode: 111,
+                message: message.into(),
+            }
+        }
 
-    #[test]
-    #[should_panic]
-    fn err_requires_nonzero() {
-        ExitError::err(0, "bogus");
-    }
+        // TODO: combine 126 and 127
```

Comment:
Did you mean to leave this TODO in here? It seems like it wouldn't be too much work to combine 126 and 127 in this PR.